src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java

```diff
@@ -269,4 +401,843 @@ public static String getParentModule(Class<?> moduleClass)
+        treeWalkerConfig.addChild(scraperCheckConfig);
+        try {
+            checker.configure(defaultConfiguration);
+            final List<File> filesToProcess = List.of(moduleFile);
+            checker.process(filesToProcess);
+            checker.destroy();
+        }
+        catch (CheckstyleException checkstyleException) {
+            final String message = String.format(Locale.ROOT, "Failed processing %s", moduleName);
+            throw new MacroExecutionException(message, checkstyleException);
+        }
+    }
+
+    /**
+     * Get a set of properties for the given class.
+     *
+     * @param clss the class to get the properties for.
+     * @return a set of properties for the given class.
+     */
+    public static Set<String> getProperties(Class<?> clss) {
+        final Set<String> result = new TreeSet<>();
+        final PropertyDescriptor[] propertyDescriptors = PropertyUtils.getPropertyDescriptors(clss);
+
+        for (PropertyDescriptor propertyDescriptor : propertyDescriptors) {
+            if (propertyDescriptor.getWriteMethod() != null) {
+                result.add(propertyDescriptor.getName());
+            }
+        }
+
+        return result;
+    }
+
+    /**
+     * Checks if the property is a global property. Global properties that are not
+     * documented.
+     *
+     * @param clss the class of the module.
+     * @param propertyName the name of the property.
+     * @return true if the property is a global property.
+     */
+    private static boolean isGlobalProperty(Class<?> clss, String propertyName) {
+        return AbstractCheck.class.isAssignableFrom(clss)
+                    && CHECK_PROPERTIES.contains(propertyName)
+                || AbstractJavadocCheck.class.isAssignableFrom(clss)
+                    && JAVADOC_CHECK_PROPERTIES.contains(propertyName)
+                || AbstractFileSetCheck.class.isAssignableFrom(clss)
+                    && FILESET_PROPERTIES.contains(propertyName);
+    }
+
+    /**
+     * Checks if the property is supposed to be documented.
+     *
+     * @param clss the class of the module.
+     * @param propertyName the name of the property.
+     * @return true if the property is supposed to be documented.
+     */
+    private static boolean isUndocumentedProperty(Class<?> clss, String propertyName) {
+        return UNDOCUMENTED_PROPERTIES.contains(clss.getSimpleName() + DOT + propertyName);
+    }
+
+    /**
+     * Gets properties that are not explicitly captured but should be documented if
+     * certain conditions are met.
+     *
+     * @param instance the instance of the module.
+     * @param clss the class of the module.
+     * @return the non explicit properties.
+     */
+    private static Set<String> getNonExplicitProperties(
+            Object instance, Class<?> clss) {
+        final Set<String> result = new TreeSet<>();
+        if (AbstractCheck.class.isAssignableFrom(clss)) {
```

Comment:
I do not understand why we must pass around `clss` and use `Class#isAssignableFrom` everywhere. This method is slower and has a different implication than the `instanceOf` keyword. I am good with this code for now, but we should improve in the future at some point.

Reply:
Not sure as well, but if it works, let's leave it. I added a draft task with a link to your comment on the project board.

Reply:
It could just be hindsight. The original tests were built out slowly, and they eventually got pretty massive. It was rare we optimized it. When I first started creating this, I think I prefer to keep the class because I actually wanted to avoid instancing it. The code is about the class structure and not its instance. However, to get the default value of the fields, we were forced to use the instance since fields could be initialized directly, through a method, or in the constructor.

I was going to say that we were using the class and you can't use instanceof with it, but it does look like we pass around the actual instance in the same corners, so we can switch.

> Maybe this is to "cheat" Intellij?

No, but we did allow more suppressions than normal in this code since it was test only and so massive.